### PR TITLE
Add simple test to be used for emitc path

### LINKED ISSUE #623: [TTNN to EmitC] Add accurate param mapping during conversion for all existing math ops
- At this moment, only `ttnn::multiply` op gets converted to a correct cpp signature.
- All `ttnn` DPS ops seem to have signature that looks like `op(arg0, arg1, optional_arg2, optional_arg3, optional_arg4, ...)`; in DPS mode, we want to populate first 2 args (if binary op) and the `optional_arg4`, which means we need to add `std::nullopt`s at call site.

### PR DESCRIPTION
**Pull request by svuckovicTT**
This PR adds an add op test for emitc path. It lives in its own folder until opt-in feature is added, as noted in [this doc](https://tenstorrent-my.sharepoint.com/:w:/r/personal/svuckovic_tenstorrent_com/_layouts/15/Doc.aspx?sourcedoc=%7Bca1db204-9282-4115-83a1-6ed46456814c%7D).

Additionally, few other tweaks were needed for this test to work:
- In TTNN to EmitC conversion
  - binary eltwise ops fix to actually call TTNN API correctly (#623)
  - renamed `Shape` to `LegalShape`
  - deallocate op converter
- Runtime
  - instead of `tensor.deallocate()` call `ttnn::operations::core:deallocate(tensor)`
  - add "init ones" method for input tensors
- Add `force` param to `DeallocateOp`, default to `false`
- Fixes to ttnn standalone

The generated code outputs the same result as when the generated flatbuffer is ran thru TTRT, with one additional fix needed in the generated code: `ttnn::empty` call for destination of the add op is generated with `ROW_MAJOR` layout but needs to be `TILE`, otherwise the result isn't correct.
Instead of fixing this today by creating a second place where layouts are fixed for binary eltwise ops (we already fix inputs to `TILE` layout), I'll let it sit as-is until the workarounds pass is added (@sdjordjevicTT @mrakitaTT something to think about). Not using destinations is another planned option that would "fix" this issue.

closes #623, #1166

### CODE REVIEW

File: runtime/lib/ttnn/operations/deletion/dealloc.cpp [hunk 1]
```diff
@@ -10,7 +11,7 @@ void run(const ::tt::target::ttnn::DeallocOp *op, ProgramContext &context) {
   ProgramTensorPool &tensorPool = context.getTensorPool();
   ::ttnn::Tensor &tensor = tensorPool.at(op->in()->global_id());
   DEBUG_ASSERT(tensor.is_allocated());
-  tensor.deallocate();
+  ttnn::operations::core::deallocate(tensor);
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/deletion/dealloc.cpp:14]: Nit: `::ttnn::deallocate(tensor);`

**Review by nsmithtt — APPROVED** [review]

File: tools/ttnn-standalone/ttnn-standalone.cpp [hunk 2]
```diff
@@ -6,53 +6,61 @@
 
 // To generate forward function, run:
 // ./build/bin/ttmlir-opt --ttir-load-system-desc --ttir-implicit-device
-// --ttir-layout --convert-ttir-to-ttnn --convert-ttnn-to-emitc
-// test/ttmlir/Dialect/TTNN/simple_multiply.mlir | ./build/bin/ttmlir-translate
+// --ttir-layout --convert-ttir-to-ttnn --ttnn-decompose-layouts
+// --ttnn-deallocate --convert-ttnn-to-emitc
+// test/ttmlir/Silicon/TTNN/emitc/simple_add.mlir | ./build/bin/ttmlir-translate
 // -mlir-to-cpp -allow-unregistered-dialect
```
Comment (rpavlovicTT) [inline tools/ttnn-standalone/ttnn-standalone.cpp:12]: is dash missing here `--mlir-to-cpp` ?

**Review by rpavlovicTT — APPROVED** [review]

File: tools/ttnn-standalone/ttnn-standalone.cpp (hunk 2 above)
Comment (svuckovicTT) [reply tools/ttnn-standalone/ttnn-standalone.cpp:12]: In the official docs, it's with a single dash: https://mlir.llvm.org/docs/Dialects/EmitC/

If you do `./build/bin/ttmlir-translate --help`, it's with two.

I'll change it to two dashes next time, to align with optimizer, don't want to trigger CI for just this.

**Comment by svuckovicTT** [comment]
@jnie-TT @kmabeeTT could I get a review on this? Should be very straightfwd for the runtime part!

File: runtime/lib/ttnn/operations/deletion/deallocate.cpp [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
+#include "ttnn/operations/core/core.hpp"
 
 namespace tt::runtime::ttnn::operations::deletion {
-void run(const ::tt::target::ttnn::DeallocOp *op, ProgramContext &context) {
+void run(const ::tt::target::ttnn::DeallocateOp *op, ProgramContext &context) {
   ProgramTensorPool &tensorPool = context.getTensorPool();
   ::ttnn::Tensor &tensor = tensorPool.at(op->in()->global_id());
   DEBUG_ASSERT(tensor.is_allocated());
-  tensor.deallocate();
+  ::ttnn::operations::core::deallocate(tensor, false);
```
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/deletion/deallocate.cpp:14]: Does ::ttnn::deallocate not work here? Also should we be reading the force flag from the op now that we've added it to the flatbuffer schema?

**Review by jnie-TT — CHANGES_REQUESTED** [review]

File: runtime/lib/ttnn/operations/deletion/deallocate.cpp (hunk 3 above)
Comment (svuckovicTT) [reply runtime/lib/ttnn/operations/deletion/deallocate.cpp:14]: Good catch on reading the flag, will fix!

No, `::ttnn::deallocate` doesn't exist in metal :/

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/deletion/deallocate.cpp:14]: Looking at the implementation it should be the same as from_device/to_device, where they also port the signature under the ttnn namespace?
```
}  // namespace core
}  // namespace operations

using operations::core::deallocate;
using operations::core::from_device;
using operations::core::reallocate;
using operations::core::squeeze_from_4D;
using operations::core::to_device;
using operations::core::unsqueeze_to_4D;

constexpr auto to_dtype = ttnn::register_operation_with_auto_launch_op<"ttnn::to_dtype", ttnn::operations::core::ToDtype>();
constexpr auto to_memory_config =
    ttnn::register_operation_with_auto_launch_op<"ttnn::to_memory_config", ttnn::operations::core::ToMemoryConfig>();
constexpr auto to_layout = ttnn::register_operation_with_auto_launch_op<"ttnn::to_layout", ttnn::operations::core::ToLayout>();

}  // namespace ttnn

```

Comment (svuckovicTT) [reply runtime/lib/ttnn/operations/deletion/deallocate.cpp:14]: Oh lol, missed that! Seems noone is using it in that format, I'll change it :)

**Review by jnie-TT — APPROVED** [review]

**Pull request merged**